src/agent/application.cpp

```diff
@@ -60,98 +60,40 @@ Application::Application(const std::string               &aInterfaceName,
 #else
     , mBackboneInterfaceName(aBackboneInterfaceNames.empty() ? "" : aBackboneInterfaceNames.front())
 #endif
+    , mHost(Ncp::ThreadController::CreateInstance(mInterfaceName.c_str(),
+                                                  aRadioUrls,
+                                                  mBackboneInterfaceName,
+                                                  /* aDryRun */ false,
+                                                  aEnableAutoAttach))
 #if OTBR_ENABLE_MDNS
     , mPublisher(Mdns::Publisher::Create([this](Mdns::Publisher::State aState) { this->HandleMdnsState(aState); }))
 #endif
 #if OTBR_ENABLE_VENDOR_SERVER
     , mVendorServer(vendor::VendorServer::newInstance(*this))
 #endif
 {
-    mHost = MakeUnique<Ncp::RcpHost>(mInterfaceName.c_str(), aRadioUrls, mBackboneInterfaceName,
-                                     /* aDryRun */ false, aEnableAutoAttach);
-
-#if OTBR_ENABLE_BORDER_AGENT
-    mBorderAgent = MakeUnique<BorderAgent>(*mHost, *mPublisher);
-#endif
-#if OTBR_ENABLE_BACKBONE_ROUTER
-    mBackboneAgent = MakeUnique<BackboneRouter::BackboneAgent>(*mHost, aInterfaceName, mBackboneInterfaceName);
-#endif
-#if OTBR_ENABLE_SRP_ADVERTISING_PROXY
-    mAdvertisingProxy = MakeUnique<AdvertisingProxy>(*mHost, *mPublisher);
-#endif
-#if OTBR_ENABLE_DNSSD_DISCOVERY_PROXY
-    mDiscoveryProxy = MakeUnique<Dnssd::DiscoveryProxy>(*mHost, *mPublisher);
-#endif
-#if OTBR_ENABLE_TREL
-    mTrelDnssd = MakeUnique<TrelDnssd::TrelDnssd>(*mHost, *mPublisher);
-#endif
-#if OTBR_ENABLE_OPENWRT
-    mUbusAgent = MakeUnique<ubus::UBusAgent>(*mHost);
-#endif
-#if OTBR_ENABLE_REST_SERVER
-    mRestWebServer = MakeUnique<rest::RestWebServer>(*mHost, aRestListenAddress, aRestListenPort);
-#endif
-#if OTBR_ENABLE_DBUS_SERVER && OTBR_ENABLE_BORDER_AGENT
-    mDBusAgent = MakeUnique<DBus::DBusAgent>(*mHost, *mPublisher);
-#endif
-
-    OT_UNUSED_VARIABLE(aRestListenAddress);
-    OT_UNUSED_VARIABLE(aRestListenPort);
+    if (mHost->GetCoprocessorType() == OT_COPROCESSOR_RCP)
+    {
+        ConstructRcpMode(aRestListenAddress, aRestListenPort);
```

Comment:
nit: to be consistent?
```suggestion
        CreateRcpMode(aRestListenAddress, aRestListenPort);
```